crates/ruff_python_formatter/src/statement/suite.rs

```diff
@@ -56,8 +58,21 @@ impl FormatRule<Suite, PyFormatContext<'_>> for FormatSuite {
                 return Ok(());
             };
 
-            // First entry has never any separator, doesn't matter which one we take.
-            write!(f, [first.format()])?;
+            if matches!(self.level, SuiteLevel::Nested)
+                && is_class_or_function_definition(first)
+                && !comments.has_leading_comments(first)
+            {
+                // Add an empty line for any nested functions or classes defined within non-function
+                // or class compound statements, e.g.:
+                // ```python
+                // if True:
+                //     def test():
+                //         ...
+                // ```
+                write!(f, [empty_line(), first.format()])?;
+            } else {
+                write!(f, [first.format()])?;
+            }
```

Comment:
Nit: I prefer to move the "common" case out of branches
```suggestion
				empty_line().fmt(f)?;
			}
			
			first.format().fmt(f)?;
```